app/src/androidTest/java/com/github/HumanLearning2021/HumanLearningApp/view/dataset_editing/CategoriesEditingActivityTest.kt

```diff
@@ -102,61 +82,68 @@ class CategoriesEditingActivityTest {
     fun rowButtonViewIsDisplayedWhenAddButtonIsClicked() {
         onView(withId(R.id.button_add)).perform(click())
         onView(withId(R.id.button_add)).check(ViewAssertions.matches(isDisplayed()))
-
-
     }
 
     @Test
     fun rowViewIsAddedWhenAddButtonIsClicked() {
         onView(withId(R.id.button_add)).perform(click())
-        waitFor(1000)
+        waitFor(1) // increase if needed
         onView(withId(R.id.parent_linear_layout)).check(
             ViewAssertions.matches(
-                hasChildCount(categories.size + 1)
+                hasChildCount(dataset.categories.size + 1)
             )
         )
     }
 
     @Test
     fun rowViewIsRemovedWhenRemoveButtonIsClicked() {
         runBlocking {
-            val delayBeforeTestStart: Long = 1000
-            waitFor(delayBeforeTestStart)
-            assumeTrue(categories.size == 1)
-            onView(withId(R.id.button_remove)).perform(click())
-            waitFor(1000)
+            waitFor(1) // increase if needed
+            val nbCategories = dataset.categories.size
+            assumeTrue(dataset.categories.isNotEmpty())
+
+            // click on first remove button
+            onView(withIndex(withId(R.id.button_remove), 0)).perform(click())
+
             onView(withId(R.id.parent_linear_layout)).check(
                 ViewAssertions.matches(
-                    hasChildCount(0)
+                    hasChildCount(nbCategories - 1)
                 )
             )
-            dataset = dbManagement.getDatasetById(datasetId)!!
-            assert(nbCategories - 1 == dataset.categories.size)
+            val updatedDataset = dbMgt.getDatasetById(dataset.id)!!
+            val actual = updatedDataset.categories.size
+            val expected = nbCategories - 1
+            assert(expected == actual) {
```

Comment:
might be better to use an hamcrest `hasSize` instead of reinventing it

Reply:
very true xD thanks